datumaro/components/project.py

```diff
@@ -2,488 +2,2181 @@
+
+        for t, s in statuses.items():
+            if s == DiffStatus.foreign_modified:
+                # TODO: compute a patch and a new stage, remove allow_foreign
+                if allow_foreign:
+                    log.warning("The source '%s' has been changed "
+                        "without Datumaro. It will be saved, but it will "
+                        "only be available for reproduction from the cache.", t)
+                else:
+                    raise ForeignChangesError(
+                        "The source '%s' is changed outside Datumaro. You can "
+                        "restore the latest source revision with 'checkout' "
+                        "command." % t)
+
+        for s in self.working_tree.sources:
+            self.refresh_source_hash(s, no_cache=no_cache)
+
+        tree_dir = osp.join(self._aux_dir, ProjectLayout.tree_dir)
+        self.working_tree.save()
+        self._git.add(tree_dir, base=tree_dir)
+        self._git.add([
+            osp.join(self._root_dir, '.dvc', '.gitignore'),
+            osp.join(self._root_dir, '.dvc', 'config'),
+            osp.join(self._root_dir, '.dvcignore'),
+            osp.join(self._root_dir, '.gitignore'),
+            osp.join(self._aux_dir, '.gitignore'),
+        ], base=self._root_dir)
+        head = self._git.commit(message)
+
+        copytree(tree_dir, self.cache_path(head))
+
+        self._head_tree = None
+
+        return head
 
     @staticmethod
-    def import_from(path, format=None, env=None, **options):
-        if env is None:
-            env = Environment()
-
-        if not format:
-            matches = env.detect_dataset(path)
-            if not matches:
-                raise NoMatchingFormatsError()
-            if 1 < len(matches):
-                raise MultipleFormatsMatchError(matches)
-            format = matches[0]
-        elif not env.is_format_known(format):
-            raise UnknownFormatError(format)
-
-        if format in env.importers:
-            project = env.make_importer(format)(path, **options)
-        elif format in env.extractors:
-            project = Project(env=env)
-            project.add_source('source', {
-                'url': path,
-                'format': format,
-                'options': options,
-            })
+    def _copy_dvc_dir(src_dir, dst_dir):
+        for name in ['config', '.gitignore']:
+            os.replace(osp.join(src_dir, name), osp.join(dst_dir, name))
+
+    def checkout(self, rev: Optional[Revision] = None,
+            sources: Union[None, str, List[str]] = None, force: bool = False):
+        """
+        Copies tree and objects from cache to working tree.
+
+        Sets HEAD to the specified revision, unless targets specified.
+        When sources specified, only copies objects from cache to working tree.
+        """
+
+        assert sources is None or isinstance(sources, (str, list)), sources
+        if sources:
+            sources = set(sources)
```

Comment:
This will likely result in unexpected behavior if `sources` has type `str`.

Reply:
Fixed